server/registry/internal/storage/delete.go

```diff
@@ -21,86 +21,171 @@ import (
 	"github.com/apigee/registry/server/registry/names"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"gorm.io/gorm"
 )
 
-func (c *Client) DeleteProject(ctx context.Context, name names.Project) error {
-	for _, model := range []interface{}{
-		models.Project{},
-		models.Api{},
-		models.Deployment{},
-		models.DeploymentRevisionTag{},
-		models.Version{},
-		models.Spec{},
-		models.SpecRevisionTag{},
-		models.Blob{},
-		models.Artifact{},
-	} {
-		op := c.db.Where("project_id = ?", name.ProjectID)
-		if err := op.Delete(model).Error; err != nil {
-			return status.Error(codes.Internal, err.Error())
+func (c *Client) DeleteProject(ctx context.Context, name names.Project, cascade bool) error {
+	err := c.db.Transaction(func(tx *gorm.DB) error {
+		var count int64
+		for _, model := range []interface{}{
+			models.Project{},
+			models.Api{},
+			models.Deployment{},
+			models.DeploymentRevisionTag{},
+			models.Version{},
+			models.Spec{},
+			models.SpecRevisionTag{},
+			models.Blob{},
+			models.Artifact{},
+		} {
+			op := tx.Where("project_id = ?", name.ProjectID)
+			if err := op.Delete(model).Error; err != nil {
+				return err
+			}
+
+			count += op.RowsAffected
 		}
-	}
 
-	return nil
+		if count > 1 && !cascade {
+			return status.Errorf(codes.FailedPrecondition, "cannot delete child resources in non-cascading mode")
```

Comment:
Nice - assuming this aborts the transaction. Can we be sure that when only one row is affected, it has to correspond to the resource at the top of the hierarchy? (I think probably yes under normal conditions, but if something gets messed up, probably not)

Reply:
Before calling these database deletes we do an existence check on the resource in the RPC handlers. Unless something happens to the project between that transaction and this one, we know it exists at this point. Then unless gorm is incorrectly reporting affected rows, we know one of the affected rows is the deleted project.

I've been thinking about keeping the database operations for each API call in a single transaction (for a later PR). In that case, the first operation in the transaction would be the existence check.

Reply:
SGTM, thanks for clarifying.